### PR TITLE
IGNITE-23291 Improve the local compaction mechanism of Metastorage

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-23291

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java [hunk 1]
```diff
@@ -17,186 +17,150 @@
...
+        testCompactRevision5();
+        testCompactRevision6();
     }
 
     @Test
-    public void testInvokeCompactionBeforeAnyEntry() {
-        byte[] key = key(0);
-        byte[] value1 = keyValue(0, 0);
-        byte[] value2 = keyValue(0, 1);
+    public void testCompactEmptyStorage() {
```
Comment (rpuch) [inline modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java:145]: Why is it about an empty storage as the storage seems to contain something (namely the data added in the 'before each' method)

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.apache.ignite.internal.metastorage.server.KeyValueStorageUtils.NOTHING_TO_COMPACT_INDEX;
+import static org.apache.ignite.internal.metastorage.server.KeyValueStorageUtils.indexToCompact;
+import static org.apache.ignite.internal.metastorage.server.KeyValueStorageUtils.toUtf8String;
+import static org.apache.ignite.internal.util.ArrayUtils.LONG_EMPTY_ARRAY;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+/** For {@link KeyValueStorageUtils} testing. */
```
Comment (rpuch) [inline modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtilsTest.java:29]: Do we need this javadoc? There is a convention that `FooTest` is for testing `Foo`, so it seems obvious and redundant

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+        assertEquals(0, indexToCompact(keyRevisions, 2, revision -> true));
+        assertEquals(NOTHING_TO_COMPACT_INDEX, indexToCompact(keyRevisions, 2, revision -> false));
+
+        assertEquals(0, indexToCompact(keyRevisions, 3, revision -> true));
+        assertEquals(NOTHING_TO_COMPACT_INDEX, indexToCompact(keyRevisions, 3, revision -> false));
+    }
+
+    @Test
+    void testIndexToCompactMultipleRevision() {
```
Comment (rpuch) [inline modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtilsTest.java:52]:
```suggestion
    void testIndexToCompactMultipleRevisions() {
```

File: modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java [hunk 4]
```diff
@@ -547,78 +557,42 @@ private boolean doRemove(byte[] key, long curRev) {
...
+     * Compacts the key, see the documentation of {@link KeyValueStorage#compact} for examples.
      *
-     * @param key A key.
-     * @param revs All revisions of a key.
-     * @param compactedKeysIdx Out parameter, revisions that need to be kept must be put here.
-     * @param compactedRevsIdx Out parameter, values that need to be kept must be put here.
-     * @param minRevisionToKeep Minimum revision that should be kept.
+     * @param key Target key.
+     * @param revs Key revisions.
+     * @param compactionRevision Revision up to which the key will be compacted.
```
Comment (rpuch) [inline modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java:564]:
```suggestion
     * @param compactionRevision Revision up to which (inclusively) the key will be compacted.
```

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorage.java [hunk 5]
```diff
@@ -233,13 +234,31 @@ boolean invoke(
...
+     * <ul>
+     *     <li>Compaction revision is {@code 1}: "foo" [3, 5], "bar" [2, 5 tomb].</li>
+     *     <li>Compaction revision is {@code 2}: "foo" [3, 5], "bar" [5 tomb].</li>
+     *     <li>Compaction revision is {@code 3}: "foo" [5], "bar" [5 tomb].</li>
+     *     <li>Compaction revision is {@code 4}: "foo" [5], "bar" [5 tomb].</li>
+     *     <li>Compaction revision is {@code 5}: "foo" [5].</li>
+     *     <li>Compaction revision is {@code 6}: "foo" [5].</li>
+     * </ul>
+     *
+     * @param revision Revision up to which the metastorage keys will be compacted.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorage.java:257]: thanks

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+     * to be compacted.
+     *
+     * <p>If the returned index points to the last revision and if the last revision is <b>not</b> a tombstone, then the returned index is
+     * decremented by 1.</p>
+     *
+     * @param keyRevisions Metastorage key revisions in ascending order.
+     * @param compactionRevisionInclusive Revision up to which you need to compact (inclusive).
+     * @param tombstoneKeyRevisionPredicate Predicate to test whether a key revision is a tombstone.
+     */
+    public static int indexToCompact(long[] keyRevisions, long compactionRevisionInclusive, LongPredicate tombstoneKeyRevisionPredicate) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java:40]: why not

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java [hunk 7]
```diff
@@ -1026,104 +1027,41 @@ private boolean addToBatchForRemoval(WriteBatch batch, byte[] key, long curRev,
...
+     * Compacts the key, see the documentation of {@link KeyValueStorage#compact} for examples.
      *
      * @param batch Write batch.
-     * @param key   Target key.
-     * @param revs  Revisions.
-     * @param minRevisionToKeep Minimum revision that should be kept.
-     * @throws RocksDBException If failed.
+     * @param key Target key.
+     * @param revs Key revisions.
+     * @param compactionRevision Revision up to which the key will be compacted.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:1035]: thanks

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtilsTest.java (hunk 2 above)
Comment (tkalkirill) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtilsTest.java:29]: Maybe, but I stick to this style, I'll leave it.

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtilsTest.java (hunk 3 above)
Comment (tkalkirill) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtilsTest.java:52]: thanks

File: modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java (hunk 4 above)
Comment (tkalkirill) [reply modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java:564]: thanks

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            i--;
+        }
+
+        return i;
+    }
+
+    private static int indexToCompact(long[] keyRevisions, long compactionRevisionInclusive) {
+        int index = NOTHING_TO_COMPACT_INDEX;
+
+        for (int i = 0; i < keyRevisions.length; i++) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java:53]: Yep, thanks!

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+     *
+     * @param keyRevisions Metastorage key revisions in ascending order.
+     * @param compactionRevisionInclusive Revision up to which you need to compact (inclusive).
+     * @param tombstoneKeyRevisionPredicate Predicate to test whether a key revision is a tombstone.
+     */
+    public static int indexToCompact(long[] keyRevisions, long compactionRevisionInclusive, LongPredicate tombstoneKeyRevisionPredicate) {
+        int i = indexToCompact(keyRevisions, compactionRevisionInclusive);
+
+        if (i != NOTHING_TO_COMPACT_INDEX && i == keyRevisions.length - 1 && !tombstoneKeyRevisionPredicate.test(keyRevisions[i])) {
+            i--;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java:44]: Yes, we shouldn't remove the last element unless it's a tombstone.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        return i;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java:47]: try to fix it

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java (hunk 1 above)
Comment (tkalkirill) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java:145]: Damn it, you're right, I'll fix it!

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java [hunk 11]
```diff
@@ -17,186 +17,150 @@
...
-        storage.put(key, value, clock.now());
-        storage.remove(key, clock.now());
+        var fooKey = new ByteArray(FOO_KEY);
+        var barKey = new ByteArray(BAR_KEY);
 
-        long lastRevision = storage.revision();
+        var iif = new If(
+                new AndCondition(new ExistenceCondition(Type.EXISTS, FOO_KEY), new ExistenceCondition(Type.EXISTS, BAR_KEY)),
+                new Statement(ops(put(fooKey, SOME_VALUE), remove(barKey)).yield()),
+                new Statement(ops(noop()).yield())
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java:68]: Probably possible, I didn’t find any examples, can you tell me how exactly to make it for storage and not for a MetastorageManager?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java [hunk 12]
```diff
@@ -1026,104 +1027,41 @@ private boolean addToBatchForRemoval(WriteBatch batch, byte[] key, long curRev,
...
 
-        if (onlyLastRevisionLeft && value.tombstone()) {
-            // We don't have any previous revisions for this entry and the single existing is a tombstone,
-            // so we can remove it from index.
-            index.delete(batch, key);
-        } else {
-            // Keeps revisions starting with idxToKeepFrom.
-            index.put(batch, key, longsToBytes(revs, idxToKeepFrom));
+            if (indexToCompact == revs.length - 1) {
+                index.delete(batch, key);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:1052]: Fix it!

[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:1052]: Add todo

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java:67]: Oh, now I see the reason. Nevermind :)

**Review by rpuch — APPROVED** [review]

**Pull request merged**